render/canvas/src/lib.rs

```diff
@@ -104,12 +104,108 @@ impl CanvasFillStyle {
     }
 }
 
+/// Stores the actual bitmap data on the browser side in one of two ways,
+/// each better suited for different scenarios and source data formats.
+enum BitmapDataStorage {
+    /// Utilizes the JPEG decoder of the browser, and can be drawn onto a canvas directly.
+    /// Needs to be drawn onto a canvas to retrieve the stored pixel data.
+    HtmlElement(HtmlImageElement),
+    /// Much easier to create from raw RGB[A] data, but needs to be "put" into a
+    /// temporary canvas before it can be drawn onto the target canvas with transforms.
+    /// The pixel data can be easily retrieved, directly from the buffer.
+    ImageData(ImageData),
+}
+
 #[allow(dead_code)]
 struct BitmapData {
-    image: HtmlImageElement,
+    image: BitmapDataStorage,
     width: u32,
     height: u32,
-    data: String,
+    /// Might be computed lazily if not available at creation.
+    data_uri: Option<String>,
```

Comment:
Could use `RefCell` here for interior mutability instead of changing the methods to `mut`. Don't think it's a big deal either way.